File: content-solutions/two-pointers/fruit-basket.js
function solution(trees, k) {
  let l = 0;
  const count = new Map();
  let best = 0;
  for (let r = 0; r < trees.length; r++) {
    count.set(trees[r], (count.get(trees[r]) ?? 0) + 1);
    while (count.size > k) {
      const left = trees[l];
      count.set(left, count.get(left) - 1);
      if (count.get(left) === 0) count.delete(left);
      l++;
    }
    best = Math.max(best, r - l + 1);
  }
  return best;
}

File: content-solutions/two-pointers/fruit-basket.java
class Solution {
    public int solution(String trees, int k) {
        int[] count = new int[26];
        int kinds = 0, l = 0, best = 0;
        for (int r = 0; r < trees.length(); r++) {
            if (count[trees.charAt(r) - 'a']++ == 0) kinds++;
            while (kinds > k) {
                int left = trees.charAt(l) - 'a';
                count[left]--;
                if (count[left] == 0) kinds--;
                l++;
            }
            best = Math.max(best, r - l + 1);
        }
        return best;
    }
}
